misc improvements to robust training example

fabianp opened the pull request:
main changes:

  * Fixes #134 by normalizing in-place.
  * Plot convergence curves for both clean and adversarial accuracy.
  * Replace the fast-sign-gradient method by the much more powerful
    PGD method.
  * Be able to select different datasets.
  * Homogeneize API wrt to the other examples. For example, this now
    uses the same load_dataset, CNN, loss_fun, accuracy than
    flax_image_classif.py . Most of the command line flags have also
    been homogeneized.

fabianp commented:
Output of example:
![Figure_1](https://user-images.githubusercontent.com/277639/146940280-e2d9c5b4-6713-453b-b446-faeefcc142a2.png)

fabianp commented:
@GeoffNN : I can't select you as official reviewer but it would be good if you can take a look before I ask for the official approval

GeoffNN commented on examples/deep_learning/plot_robust_training.py:
```diff
@@ -0,0 +1,229 @@
+  params = net.init(rng, jnp.zeros(input_shape))["params"]
+
+  state = solver.init_state(params)
+  rng = jax.random.PRNGKey(0)
```
Would it be better to split the rngs, rather than initialize the same one multiple times? This seems bug prone if someone reuses the code as a template.

GeoffNN reviewed (commented):
A few cosmetic comments, other than that looks good!

GeoffNN commented:
One last comment: it would be good to include test set accuracy, both for the clean images and adversarial images.

GeoffNN commented on examples/deep_learning/plot_robust_training.py:
```diff
@@ -0,0 +1,229 @@
+flags.DEFINE_enum("dataset", "mnist", dataset_names, "Dataset to train on.")
+flags.DEFINE_integer("epochs", 5, "Number of passes over the dataset.")
+flags.DEFINE_integer("train_batch_size", 256, "Batch size at train time.")
+flags.DEFINE_integer("test_batch_size", 1024, "Batch size at test time.")
```
`test_batch_size` isn't used

fabianp commented:
Thanks a lot @GeoffNN ! super useful comments. Do you mind taking another look before I ask LGTM? thanks!

GeoffNN commented on examples/deep_learning/plot_robust_training.py:
```diff
@@ -0,0 +1,253 @@
+flags.DEFINE_float("learning_rate", 0.001, "Learning rate.")
+flags.DEFINE_float(
+    "epsilon", 0.01,
+    "Adversarial perturbations will be constrained to lie the L-infinity ball of radius epsilon."
```
to lie in*

GeoffNN commented on examples/deep_learning/plot_robust_training.py:
```diff
@@ -0,0 +1,253 @@
+with respect to the projected gradient descent (PGD) method.
+
+The Projected Gradient Descent Method (PGD) is a simple yet effective method to
+generate adversarial images. At each iteration, adds a small perturbation
```
it adds* ?

fabianp replied:
done

GeoffNN commented on examples/deep_learning/plot_robust_training.py:
```diff
@@ -0,0 +1,253 @@
+  adversarial_accuracy_test = []
+  for it in range(solver.maxiter):
+    # Training
+    images, labels = next(train_ds)
```
What happens here when `it` is larger than `len(dataset)` does it cycle back through, or does it reshuffle ?

GeoffNN replied:
It might be better to write this in `epoch' logic, and sample a new permutation over the dataset at each epoch. Then we want to test over the test set at the end of 1 training epoch: right now, we only have the test accuracies for single test batches vs over the whole test set.

GeoffNN reviewed (changes requested):
Looks almost good to me! One thing: the testing shouldn't take place within the training loop. We should have a loop over epochs, with 2 subloops: one training loop corresponding to a sampled permutation over the train data, and a testing loop which computes the test accuracy over the whole test set at each epoch.

fabianp commented:
Thanks @GeoffNN for the review, I think I addressed all points! I thin its ready for review @mblondel 

this is what the current output looks like:

![Figure_1](https://user-images.githubusercontent.com/277639/152017040-37ad1a18-3011-4067-9219-88edf95220f9.png)

mblondel commented on examples/deep_learning/plot_robust_training.py:
```diff
@@ -0,0 +1,268 @@
+
+    for _ in range(maxiter):
+      # compute gradient of the loss wrt to the image
+      grad = jax.grad(adversarial_loss)(image_perturbation)
```
It would probably be better to store `grad_fun = jax.grad(adversarial_loss)` outside of the loop, though `jax.jit` may be able to optimize that away.

fabianp replied:
good point, though I'll call it grad_adversarial since its the grad of the adversarial loss (wrt to images )

mblondel commented on examples/deep_learning/plot_robust_training.py:
```diff
@@ -0,0 +1,268 @@
+generate adversarial images. At each iteration, it adds a small perturbation
+in the direction of the sign of the gradient with respect to the input followed
+by a projection onto the infinity ball. The gradient sign ensures this
+perturbation locally maximizes the objective, while projection ensures this
```
```suggestion
perturbation locally maximizes the objective, while the projection ensures this
```

mblondel reviewed (approved):
LGTM, apart from my two minor comments. Thanks @fabianp!

fabianp commented:
addresed @mblondel's comments

fabianp commented:
@mblondel : changed the changelog, do you mind taking a quick look at that before merging?
